elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,248 @@
+
+==== Expression Resolver
+
+Presently WildFly Core supports an expression resolver that can delegate to a Vault configuration and if that is not available fall back to use either system properties or environment variables.  This will be updated to make use of the `CapabilityRegistry` and attempt to lookup a capability using a pre-defined constant (To be defined) which exposes an expression resolver runtime API.  The expression resolver looked up using a capability will be used first to attempt to resolve any expression, if this is successful the resolved value will be used - if not we will fallback to the existing behaviour of using the vault and if that fails using a system property or environment variable.
+
+A new singleton resource will be added to the elytron subsystem called `expression=encryption` the purpose of this resource is to contain the configuration both for the encryption and decryption of values.  As a singleton we know only one instance of this resource can be defined within the subsystem reducing some of the complexity multiple instances would cause.  The resource will contain the definition of one or more resolvers to decrypt the inline expressions.
+
+The format of the expressions resolved using the WildFly Elytron expression resolver will be `${ENC:Resolver:ENCRYPTED_DATA}` where `Resolver` is a reference to the specific resolver defined in the `expression=encryption` resource and `ENCRYPTED_DATA` is the data to be decrypted encoded using Base64.
+
+The `expression=encryption` resource will register it's expression resolver capability with the `CapabilityRegistry` making it available for runtime resolution of expressions.
+
+The `expression=encryption` resource will contain three attributes, the first attribute being `resolvers` which will be used to define one or more resolvers to handle the encryption and decryption of values.  Each resolver will support the following configuration: -
+
+ * Reference to a credential store and alias to identify a specific previously created `SecretKey` to use.
+ * Name of the algorithm to use for the encryption / decryption.
+ * Any additional parameters such as `salt`, `iteration count` etc.. as required to configure the algorithm.
+
+The second attribute on the `expression=encryption` resource will be `default-resolver` which will be used to optionally specify which resolver is the default, the expression can then be simplified to `${ENC:ENCRYPTED_DATA}`.
+
+As expressions are already widely in use with the application server there is a small possibility that users may have already defined expressions with a prefix of `ENC:` expecting it to be resolved as a system property.  The `expression=encryption` resource will also have a third attribute `prefix` attribute which will allow an alternative prefix to `ENC` to be specified.
```

Comment:
What if users will already have defined `${ENC}` variable in their configuration. Won't this cause a backward compatibility issues when the `${ENC}` variable will start causing errors? Customer must be aware of fact, that this might happen.

This also leads me to following questions:

1. Will be the `expression=encryption` resource defined by default?
1. What will happen if I set the `prefix` to another value with expressions already defined? Do I have to as a user manually update all my expression prefixes?

Reply:
The reason we will make the prefix configurable is so they can select a prefix not already in use, it is unlikely they will have something that starts 'ENC' but not impossible so if they are using this new expression resolution they will be able to override that.

The expression=encryption resource will not be included in the default configuration as we do not ship a default credential store so there will be no compatibility issue just by moving to a new version.

Reply:
Thank you for response! Though you probably forgot to respond to question number 2.

Reply:
Regarding '2', they MUST use a prefix that is not already in use in the model - if they specify a prefix already used in their expressions then their expression will be incorrectly picked up by the expression=encryption expression resolver.  In that situation it is safer to fail with an error.

We could do something where we silently log failures and continue allowing a fallback but those can become a pain to debug where we have allowed a single expression through multiple resolvers that it matches.